SimTracker/SiPhase2Digitizer/plugins/Pixel3DDigitizerAlgorithm.cc

```diff
@@ -0,0 +1,452 @@
+//#include "CondFormats/SiPixelObjects/interface/GlobalPixel.h"
+#include "CondFormats/DataRecord/interface/SiPixelQualityRcd.h"
+#include "CondFormats/DataRecord/interface/SiPixelFedCablingMapRcd.h"
+#include "CondFormats/DataRecord/interface/SiPixelLorentzAngleSimRcd.h"
+#include "CondFormats/SiPixelObjects/interface/SiPixelFedCablingMap.h"
+//#include "CondFormats/SiPixelObjects/interface/SiPixelFedCablingTree.h"
+//#include "CondFormats/SiPixelObjects/interface/SiPixelFedCabling.h"
+//#include "CondFormats/SiPixelObjects/interface/PixelIndices.h"
+//#include "CondFormats/SiPixelObjects/interface/SiPixelLorentzAngle.h"
+//#include "CondFormats/SiPixelObjects/interface/SiPixelQuality.h"
+//#include "CondFormats/SiPixelObjects/interface/PixelROC.h"
+//#include "CondFormats/SiPixelObjects/interface/LocalPixel.h"
+//#include "CondFormats/SiPixelObjects/interface/CablingPathToDetUnit.h"
+//#include "CondFormats/SiPixelObjects/interface/SiPixelFrameReverter.h"
+//#include "CondFormats/SiPixelObjects/interface/PixelFEDCabling.h"
+//#include "CondFormats/SiPixelObjects/interface/PixelFEDLink.h"
+//#include "CondFormats/SiPixelObjects/interface/PixelROC.h"
+
+// Geometry
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
+//#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+//#include "Geometry/CommonTopologies/interface/PixelTopology.h"
+//#include "Geometry/Records/interface/IdealGeometryRecord.h"
+
+//#include <iostream>
+#include <cmath>
+#include <vector>
+#include <algorithm>
+
+using namespace sipixelobjects;
+
+// REMEMBER CMS conventions:
+// -- Energy: GeV
+// -- momentum: GeV/c
+// -- mass: GeV/c^2
+// -- Distance, position: cm
+// -- Time: ns
+// -- Angles: radian
+// Some constants
+const double SPEED_OF_LIGHT = 30.0;
+const double unit_um = 1e-4;
+
+void Pixel3DDigitizerAlgorithm::init(const edm::EventSetup& es) {
+  // XXX: Just copied from PixelDigitizer Algorithm
+  //      CHECK if all these is needed
+
+  if (use_ineff_from_db_) {
+    // load gain calibration service fromdb...
+    theSiPixelGainCalibrationService_->setESObjects(es);
+  }
+
+  if (use_deadmodule_DB_) {
+    es.get<SiPixelQualityRcd>().get(SiPixelBadModule_);
+  }
+
+  if (use_LorentzAngle_DB_) {
+    // Get Lorentz angle from DB record
+    es.get<SiPixelLorentzAngleSimRcd>().get(SiPixelLorentzAngle_);
+  }
+
+  // gets the map and geometry from the DB (to kill ROCs)
+  es.get<SiPixelFedCablingMapRcd>().get(fedCablingMap_);
+  es.get<TrackerDigiGeometryRecord>().get(geom_);
+}
+
+Pixel3DDigitizerAlgorithm::Pixel3DDigitizerAlgorithm(const edm::ParameterSet& conf)
+    : Phase2TrackerDigitizerAlgorithm(conf.getParameter<edm::ParameterSet>("AlgorithmCommon"),
+                                      conf.getParameter<edm::ParameterSet>("Pixel3DDigitizerAlgorithm")),
+      // The size of the column np-junction (XXX: to be included via config)
```

Comment:
Indeed, the code is in development, the implementation of the pixels 3D is not finished yet. We decided to go for the PR because there was too many people working in private branches and it was becoming complicated to keep developments updated. We learnt about the way MDT is developing and we try to mimic. So, the point is to include the classes, geometry and workflows in the official repo and from there we can go on with the developments.